src/set_theory/surreal/dyadic.lean

```diff
@@ -178,33 +178,56 @@ begin
     linarith },
 end
 
-/-- The map `dyadic_map` sends ⟦⟨m, 2^n⟩⟧ to m • half ^ n. -/
-def dyadic_map (x : localization (submonoid.powers (2 : ℤ))) : surreal :=
-localization.lift_on x (λ x y, x • pow_half (submonoid.log y)) $
-begin
-  intros m₁ m₂ n₁ n₂ h₁,
-  obtain ⟨⟨n₃, y₃, hn₃⟩, h₂⟩ := localization.r_iff_exists.mp h₁,
-  simp only [subtype.coe_mk, mul_eq_mul_right_iff] at h₂,
-  cases h₂,
-  { simp only,
-    obtain ⟨a₁, ha₁⟩ := n₁.prop,
-    obtain ⟨a₂, ha₂⟩ := n₂.prop,
-    have hn₁ : n₁ = submonoid.pow 2 a₁ := subtype.ext ha₁.symm,
-    have hn₂ : n₂ = submonoid.pow 2 a₂ := subtype.ext ha₂.symm,
+/-- The additive monoid morphism `dyadic_map` sends ⟦⟨m, 2^n⟩⟧ to m • half ^ n. -/
```

Comment:
Can you add a lemma that makes this statement?

Reply:
I managed to get these. I'm not sure about the names at all...

```lean
lemma dyadic_map_apply (m : ℤ) (p : submonoid.powers (2 : ℤ)) : 
  dyadic_map (localization.mk m p) = m • pow_half (submonoid.log p) := rfl

lemma dyadic_map_apply' (m : ℤ) (n : ℕ) : 
  dyadic_map (localization.mk m (submonoid.pow 2 n)) = m • pow_half n :=
by rw [dyadic_map_apply, @submonoid.log_pow_int_eq_self 2 one_lt_two]
```

Reply:
Here's one more slightly different variant of the same lemma
```lean
lemma dyadic_map_apply' (m : ℤ) (n : ℕ) : 
  dyadic_map (localization.mk m ⟨2 ^ n, n, rfl⟩) = m • pow_half n :=
by rw [← submonoid.pow_apply, dyadic_map_apply, @submonoid.log_pow_int_eq_self 2 one_lt_two]
```